modeshape-jcr/src/main/java/org/modeshape/jcr/cache/document/WritableSessionCache.java

```diff
@@ -144,11 +147,19 @@ public WritableSessionCache( ExecutionContext context,
         this.referrerChangesForRemovedNodes = new HashMap<NodeKey, ReferrerChanges>();
         this.binaryReferencesByNodeKey = new ConcurrentHashMap<NodeKey, Set<BinaryKey>>();
         this.txns = repositoryEnvironment.getTransactions();
+        this.nodeTypes = repositoryEnvironment.nodeTypes();
     }
 
     protected final void assertInSession( SessionNode node ) {
         assert this.changedNodes.get(node.getKey()) == node : "Node " + node.getKey() + " is not in this session";
     }
+    
+    protected NodeTypes nodeTypes() {
+        if (this.nodeTypes == null && workspaceCache().repositoryEnvironment() != null) {
+            this.nodeTypes = workspaceCache().repositoryEnvironment().nodeTypes();
+        }  
```

Comment:
Right. But that's why I'm suggesting a brief comment, like `// this is idempotent` so that someone reviewing the code in a year can easily see that this logic still is `@ThreadSafe`.